### PR TITLE
[AU4] Clip copy&paste

### PR DESCRIPTION
**Pull request by grliszas14**
Resolves: https://github.com/audacity/audacity/issues/6766

### CODE REVIEW

File: au4/src/au3wrap/internal/processinginteraction.cpp [hunk 1]
```diff
@@ -61,6 +67,116 @@ bool ProcessingInteraction::changeClipTitle(const processing::ClipKey& clipKey,
...
+
+    // paste
+    for (const auto& data : s_clipboard.data) {
+        dstWaveTrack->Paste(begin, *data.track);
+        int numClips = dstWaveTrack->GetNumClips();
+
+        processing::ProcessingProjectPtr prj = globalContext()->currentProcessingProject();
+        for (int i = 0; i < numClips; ++i) {
+            std::shared_ptr<WaveClip> clip = DomAccessor::findWaveClip(dstWaveTrack, i);
+            prj->onClipAdded(DomConverter::clip(dstWaveTrack, clip.get(), i));
```
Comment (Eism) [inline au4/src/au3wrap/internal/processinginteraction.cpp:125]: I think we should only send clips that were added, not all clips from the current track

File: au4/src/au3wrap/internal/au3project.cpp [hunk 2]
```diff
@@ -87,6 +88,8 @@ bool Au3Project::load(const muse::io::path_t& filePath)
     TrackList& tracks = TrackList::Get(m_data->projectRef());
     for (auto pTrack : tracks) {
         pTrack->LinkConsistencyFix();
+        //! TODO AU4: adjust to actual project tempo
+        DoProjectTempoChange(*pTrack, 120);
```
Comment (Eism) [inline au4/src/au3wrap/internal/au3project.cpp:92]: Could you please explain these changes?

Comment (grliszas14) [reply au4/src/au3wrap/internal/au3project.cpp:92]: Project needs to have a tempo setup - if it doesn't when you try to copy&paste a clip then Paste() function throws an exception. So for now I've set a hardcoded 120 BPM (it's not used anywhere yet except copy&paste logic)

File: au4/src/au3wrap/internal/processinginteraction.cpp (hunk 1 above)
Comment (grliszas14) [reply au4/src/au3wrap/internal/processinginteraction.cpp]: Fixed

File: au4/src/trackedit/itrackeditinteraction.h [hunk 3]
```diff
@@ -26,6 +27,11 @@ class ITrackeditInteraction : MODULE_EXPORT_INTERFACE
     virtual muse::async::Channel<ClipKey, double /*newStartTime*/, bool /*completed*/> clipStartTimeChanged() const = 0;
 
     virtual bool changeClipTitle(const ClipKey& clipKey, const muse::String& newTitle) = 0;
+    virtual void clearClipboard() = 0;
+    virtual bool paste(double begin, TrackId trackId) = 0;
```
Comment (vsverchinsky) [inline au4/src/trackedit/itrackeditinteraction.h:31]: Won't harm to mention the clipboard in the names of the added functions

**Review by vsverchinsky — APPROVED** [review]

File: au4/src/trackedit/itrackeditinteraction.h (hunk 3 above)
Comment (grliszas14) [reply au4/src/trackedit/itrackeditinteraction.h]: fixed

**Pull request merged**